Add alias to catalog

c-herrewijn opened the pull request:
This PR adds alias information to the output of function `duckdb_functions()`. This makes it possible to query aliases which is relevant when generating documentation.

### example usage
```sql
D select distinct function_name, alias_of, function_type from duckdb_functions() where alias_of in ['arg_min', 'list_contains'] order by alias_of;
┌────────────────┬───────────────┬───────────────┐
│ function_name  │   alias_of    │ function_type │
│    varchar     │    varchar    │    varchar    │
├────────────────┼───────────────┼───────────────┤
│ min_by         │ arg_min       │ aggregate     │
│ argmin         │ arg_min       │ aggregate     │
│ array_contains │ list_contains │ scalar        │
│ array_has      │ list_contains │ scalar        │
│ list_has       │ list_contains │ scalar        │
└────────────────┴───────────────┴───────────────┘
```

Note that the relation is one directional:
```sql
D select distinct function_name, alias_of from duckdb_functions() where function_name in ['arg_min', 'argmin', 'min_by'] order by function_name;
┌───────────────┬──────────┐
│ function_name │ alias_of │
│    varchar    │ varchar  │
├───────────────┼──────────┤
│ arg_min       │ NULL     │
│ argmin        │ arg_min  │
│ min_by        │ arg_min  │
└───────────────┴──────────┘
```

Query to find aliases, regardless of which name is "canonical":
```sql
select distinct
    f2.function_name as 'aliases excl self'
from
    duckdb_functions() f1
    join duckdb_functions() f2 on (
        f1.alias_of = f2.function_name
        or f1.function_name = f2.alias_of
        or (f1.alias_of = f2.alias_of and f1.function_name != f2.function_name)
    )
where
    f1.function_name = 'argmin';
``` 

canonical functions can easily be queried with:
```sql
FROM duckdb_functions() WHERE alias_of is NULL;
```

### scope
- Column `alias_of` is filled in cases where the alias relation is explicit in the generated function headers via `using ALIAS`. This applies for the scalar and aggregate functions in:
  - `src/function/function_list.cpp`
  - `extension/core_functions/function_list.cpp`

### not in scope
- implicit, or partial aliases. For example, functions`list_extract` and `array_extract` are not considered aliases, as they don't have the exact same set of overloads. (`array_extract` is more widely applicable as it can also operate on structs)
- table functions, marcros, and pragma. Column 'alias_of' will be NULL.
- functions added or altered by extensions other then 'core_functions'. Even if some extensions might add aliases, this information is not explicit. Column 'alias_of' will be NULL.
  - see: `src/include/duckdb/main/extension_entries.hpp`
  - note that if we were to support the `alias_of` attribute for extensions, this would probably impact:
    - script: `scripts/generate_extensions_function.py`
    - functions: `BuiltinFunctions::RegisterExtensionOverloads()`, `BuiltinFunctions::AddExtensionFunction()` and the different overrides of `CatalogEntry::AlterEntry()`
    - structs: `CreateScalarFunctionInfo`, `ExtensionFunctionEntry`, `ExtensionFunctionOverloadEntry`

c-herrewijn committed:
add canonical function name

c-herrewijn committed:
Merge branch 'duckdb:main' into add_alias_to_catalog

c-herrewijn committed:
reverted option to change canonical_name via AlterEntry() as canonical_name currently is only supported for core/internal functions

Maxxen commented:
Looks good! Only feedback is that `canonical_function_name` is ... very long and kind of unwieldy to write, so id maybe prefer e.g. `alias_of` or something. But its not really up to me and I could live with the current naming.

c-herrewijn commented:
> Looks good! Only feedback is that `canonical_function_name` is ... very long and kind of unwieldy to write, so id maybe prefer e.g. `alias_of` or something. But its not really up to me and I could live with the current naming.

maybe `canonical_name`? `alias_of` might be a bit strange if you refer to yourself (both columns having the same name)

c-herrewijn committed:
rename 'canonical_name' to 'alias_of'

c-herrewijn committed:
don't print own function_name in column 'alias_of'

c-herrewijn commented:
Changed the name to `alias_of`, also changed `duckdb_functions()` so functions don't repeat their own name as alias.
New output:

```sql
D select distinct function_name, alias_of, function_type from duckdb_functions() where alias_of in ['arg_min', 'list_contains'] order by alias_of;
┌────────────────┬───────────────┬───────────────┐
│ function_name  │   alias_of    │ function_type │
│    varchar     │    varchar    │    varchar    │
├────────────────┼───────────────┼───────────────┤
│ min_by         │ arg_min       │ aggregate     │
│ argmin         │ arg_min       │ aggregate     │
│ array_contains │ list_contains │ scalar        │
│ array_has      │ list_contains │ scalar        │
│ list_has       │ list_contains │ scalar        │
└────────────────┴───────────────┴───────────────┘
```

for reference, the old version had this output
```sql
D select distinct function_name, canonical_function_name, function_type from duckdb_functions() where canonical_function_name in ['arg_min', 'list_contains'] order by canonical_function_name;
┌────────────────┬─────────────────────────┬───────────────┐
│ function_name  │ canonical_function_name │ function_type │
│    varchar     │         varchar         │    varchar    │
├────────────────┼─────────────────────────┼───────────────┤
│ argmin         │ arg_min                 │ aggregate     │
│ arg_min        │ arg_min                 │ aggregate     │
│ min_by         │ arg_min                 │ aggregate     │
│ array_has      │ list_contains           │ scalar        │
│ array_contains │ list_contains           │ scalar        │
│ list_has       │ list_contains           │ scalar        │
│ list_contains  │ list_contains           │ scalar        │
└────────────────┴─────────────────────────┴───────────────┘
```

c-herrewijn committed:
consistently use NULL values in case there is no 'alias_of'

Mytherin commented:
Thanks!
